### ISSUE TITLE
Implement preemptible regular IRQs on ARC

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#630]
**_Reported by Andrew Boie:_**

As an application developer, I would like support for nested regular interrupts on ARC so that I can have more fine grained control of interrupt preemption (and therefore interrupt latency) on ARC CPU designs that have more than 2 interrupt priority levels.

Original ARC port was targeted at Arduino 101 which just has 2 priority levels: FIRQ and regular IRQ. However, other ARC designs (em9d, em11d come to mind) which  have multiple regular IRQ priority levels. Currently the build system throws an error:

```

#if CONFIG_NUM_IRQ_PRIO_LEVELS > 2
#error "NUM_IRQ_PRIO_LEVELS>2 is not supported."
/*
 * Nesting of Regularing interrupts is not yet supported.
 * If your SOC supports more than 2, set this value to 2.
 */
#endif
```

(Imported from Jira ZEP-713)

**Comment by nashif** [comment]
**_by Andrew Boie:_**

add watchers

**Comment by nashif** [comment]
**_by Anas Nashif:_**

Ruud, can you please determine priority of this and who should be doing it? thanks.

**Comment by nashif** [comment]
**_by Francois Bedard:_**

Hi Anas,

Chuck Jordan will do this for release 1.6.

**Comment by nashif** [comment]
**_by Mark Linkmeyer:_**

_Ruud Derwig_ , can you please verify if this has already been merged and if so, please move this issue to the Merged state?  Thx.

**Comment by nashif** [comment]
**_by Mark Linkmeyer:_**

_Chuck Jordan_ , what's the status of this story?

**Comment by nashif** [comment]
**_by Francois Bedard:_**

Hi Mark,

Pull request has been submitted at https://github.com/zephyrproject-rtos/zephyr/pull/761 

Originator Wayne Ren is now out for a couple of weeks on parental leave so I'll ask Chuck to look at these changes again this week as he had some concerns.  Will try to update pull request by end of this week or early next week.  Thanks,

**Comment by nashif** [comment]
**_by Chuck Jordan:_**

I've been busy with other things and haven't gotten to this yet.
I'm concerned about interrupt latency.
I have a little toy application that drives an oled display with SPI or I2c.
The interrupt latency is crucial here to drive the display properly.
I want to try that.
For ARC targets, we want FIRQ to have a very small interrupt latency -- so unifying to one stack, and having common code for all types of interrupts, I believe, will be too slow.
Nested interrupt support is only needed for RIRQ. I would leave FIRQ alone so that FIRQ could be faster.

**Comment by nashif** [comment]
**_by Ruud Derwig:_**

Enhancement was verified and merged for 1.9.
